File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/render/VulkanOwnedTextureFrame.java
package org.maplibre.nativeffi.render;

/** Borrowed Vulkan texture frame valid only during the frame callback. */
public final class VulkanOwnedTextureFrame {
  private final FrameScope scope;
  private final long generation;
  private final int width;
  private final int height;
  private final double scaleFactor;
  private final long frameId;
  private final NativePointer image;
  private final NativePointer imageView;
  private final NativePointer device;
  private final int format;
  private final int layout;

  VulkanOwnedTextureFrame(
      FrameScope scope,
      long generation,
      int width,
      int height,
      double scaleFactor,
      long frameId,
      NativePointer image,
      NativePointer imageView,
      NativePointer device,
      int format,
      int layout) {
    this.scope = scope;
    this.generation = generation;
    this.width = width;
    this.height = height;
    this.scaleFactor = scaleFactor;
    this.frameId = frameId;
    this.image = image;
    this.imageView = imageView;
    this.device = device;
    this.format = format;
    this.layout = layout;
  }

  public long generation() {
    scope.ensureActive();
    return generation;
  }

  public int width() {
    scope.ensureActive();
    return width;
  }

  public int height() {
    scope.ensureActive();
    return height;
  }

  public double scaleFactor() {
    scope.ensureActive();
    return scaleFactor;
  }

  public long frameId() {
    scope.ensureActive();
    return frameId;
  }

  public NativePointer image() {
    scope.ensureActive();
    return image;
  }

  public NativePointer imageView() {
    scope.ensureActive();
    return imageView;
  }

  public NativePointer device() {
    scope.ensureActive();
    return device;
  }

  public int format() {
    scope.ensureActive();
    return format;
  }

  public int layout() {
    scope.ensureActive();
    return layout;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/render/VulkanOwnedTextureFrame.kt
package org.maplibre.nativeffi.render

/** Borrowed Vulkan texture frame valid only while its frame handle is open. */
public class VulkanOwnedTextureFrame
internal constructor(
  private val scope: FrameScope,
  private val generationValue: Long,
  private val widthValue: Int,
  private val heightValue: Int,
  private val scaleFactorValue: Double,
  private val frameIdValue: Long,
  private val imageValue: VulkanHandle,
  private val imageViewValue: VulkanHandle,
  private val deviceValue: NativePointer,
  private val formatValue: Int,
  private val layoutValue: Int,
) {
  /** Session generation preserved as a native `uint64_t` bit pattern in [Long]. */
  public fun generation(): Long = checked { generationValue }

  public fun width(): Int = checked { widthValue }

  public fun height(): Int = checked { heightValue }

  public fun scaleFactor(): Double = checked { scaleFactorValue }

  /** Opaque frame identity preserved as a native `uint64_t` bit pattern in [Long]. */
  public fun frameId(): Long = checked { frameIdValue }

  public fun image(): VulkanHandle = checked { imageValue }

  public fun imageView(): VulkanHandle = checked { imageViewValue }

  public fun device(): NativePointer = checked { deviceValue }

  public fun format(): Int = checked { formatValue }

  public fun layout(): Int = checked { layoutValue }

  private inline fun <T> checked(block: () -> T): T {
    scope.ensureActive()
    return block()
  }
}
